plugins/goodix-fp/fu-goodixfp-device.c

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright (C) 2016-2017 Richard Hughes <richard@hughsie.com>
+ * Copyright (C) 2020 boger wang <boger@goodix.com>
+ * 
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-chunk.h"
+#include "fu-goodixfp-common.h"
+#include "fu-goodixfp-device.h"
+
+struct _FuGoodixFpDevice {
+	FuUsbDevice	parent_instance;
+};
+
+G_DEFINE_TYPE (FuGoodixFpDevice, fu_goodixfp_device, FU_TYPE_USB_DEVICE)
+
+#define GX_USB_BULK_EP_IN  (3 | 0x80)
+#define GX_USB_BULK_EP_OUT (1 | 0x00)
+
+#define GX_USB_BULK_EP_SIZE 64
+#define GX_USB_INTERFACE    0 
+
+#define GX_USB_DATAIN_TIMEOUT        2000  /* ms */
+#define GX_USB_DATAOUT_TIMEOUT       200   /* ms */
+#define GX_FLASH_TRANSFER_BLOCK_SIZE 1000  /* 1000 */
+
+static gboolean
+goodixfp_device_cmd_send (GUsbDevice *usbdevice,
+			  guint8 cmd0,
+			  guint8 cmd1,
+			  guint8 pkg_eop,
+			  GByteArray *request,
+			  GError **error)
+{
+	gboolean ret = FALSE;
+	GxfpPkgHeader header = {0,};
+	guint32 crc_calc = 0;
+	gsize actual_len = 0;
+
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	fu_goodixfp_build_header (&header, request->len, cmd0, cmd1, pkg_eop);
+	g_byte_array_append (buf, (guint8 *)&header, sizeof(header));
+	g_byte_array_append (buf, request->data, request->len);
+	fu_goodixfp_crc32 (buf->data, sizeof(header) + request->len, &crc_calc);
+
+	fu_byte_array_append_uint32 (buf, crc_calc, G_LITTLE_ENDIAN);
+
+	/* send zero length package */
+	ret = g_usb_device_bulk_transfer (usbdevice,
+					  GX_USB_BULK_EP_OUT,
+					  NULL,
+					  0,
+					  NULL,
+					  GX_USB_DATAOUT_TIMEOUT, NULL, error);
+	if (!ret) {
+		g_prefix_error (error, "failed to request: ");
+		return FALSE;
+	}
+
+	if (g_getenv ("FWUPD_GOODIXFP_VERBOSE") != NULL) {
+		fu_common_dump_full (G_LOG_DOMAIN, "REQST",
+				     buf->data, buf->len, 16,
+				     FU_DUMP_FLAGS_SHOW_ADDRESSES);
+	}
+	/* send data */
+	ret = g_usb_device_bulk_transfer (usbdevice,
+					  GX_USB_BULK_EP_OUT,
+					  buf->data,
+					  buf->len,
+					  &actual_len,
+					  GX_USB_DATAOUT_TIMEOUT, NULL, error);
+	if (!ret) {
+		g_prefix_error (error, "failed to request: ");
+		return FALSE;
+	}
+
+	if (actual_len != buf->len) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "Invalid length");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+goodixfp_device_cmd_recv (GUsbDevice *usbdevice,
+			  GxfpCmdResp *presponse,
+			  gboolean data_reply,
+			  GError **error)
+{
+	gboolean ret = FALSE;
+	GxfpPkgHeader header = {0,};
+	guint32 crc32_calc = 0;
+	gsize actual_len = 0;
+	gsize offset = 0;
+
+	g_assert(presponse != NULL);
+
+	/*
+	* package format
+	* | zlp | ack | zlp | data |
+	*/
+	while (1)
+	{
```

Comment:
brace again

Reply:
I thought only  `if(foo) {`, ok,  i will fix it .

Reply:
`if (foo) {` actually :)